src/passes/OptimizeInstructions.cpp

```diff
@@ -1170,6 +1177,39 @@ struct OptimizeInstructions
   // Information about our locals
   std::vector<LocalInfo> localInfo;
 
+  // Check if two inputs to an instruction are equal, and can be removed. This
+  // is simpler than the general question of equality because we know they are
+  // inputs to the same instruction, and so they execute one after the other,
+  // with nothing else in the middle. So all we need to check is that they are
```

Comment:
Not sure if I understand... How do we guarantee that `left` and `right` are the inputs to the same instruction? And why do they need to e inputs from the same instruction and execute one after the other in the first place?

Reply:
We don't guarantee that here, it's a precondition for calling this method.

Basically this is called from simple peephole optimization rules, checking if two inputs to an instruction are identical, and if so, it can be simplified in various ways.

Does that make sense? If not I can try to rephrase the comment?

Reply:
I see, but I'm still not sure why it should be a precondition that they should be inputs to the same instruction and there shouldn't be any instructions in between. I get they are currently used only in those context though.